llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1447,6 +1484,388 @@ LogicalResult TopkOp::getResultTilePosition(
+    dimAndTileMapping[dimsToBlock[i]] = tiles[i];
+  return dimAndTileMapping;
+}
+
+// Return a mapping from positions `dims_pos` to their `int64_t` tile factors.
+// If the tile factor is dynamic a sentinel is inserted in the map.
+DenseMap<int64_t, int64_t> PackOp::getDimAndStaticTileMapping() {
+  SmallVector<Value> dynamicTiles;
+  SmallVector<int64_t> staticTiles;
+  dispatchIndexOpFoldResults(getMixedTiles(), dynamicTiles, staticTiles,
+                             ShapedType::kDynamicSize);
+  SmallVector<int64_t> indicesOfBlockedDims =
+      extractFromI64ArrayAttr(getDimsPos());
+
+  assert(staticTiles.size() == indicesOfBlockedDims.size() &&
+         "tiles must match indices of dimension to block");
+  // bind the dimension with the tile factor.
+  DenseMap<int64_t, int64_t> tileAndPosMapping;
+  for (auto i : llvm::seq<int64_t>(0, indicesOfBlockedDims.size()))
+    tileAndPosMapping[indicesOfBlockedDims[i]] = staticTiles[i];
+  return tileAndPosMapping;
+}
+
+// Implements `getIterationDomain` from the tiling interface. In each
+// loop the lower bound is zero and the step is one. For upper bound
+// is inferred from the output tensor.
+SmallVector<Range> PackOp::getIterationDomain(OpBuilder &builder) {
+  int64_t outputRank = getOutputRank();
+  SmallVector<Range> loopBounds(outputRank);
+  Location loc = getLoc();
+  Value zero = builder.create<arith::ConstantIndexOp>(loc, 0);
+  Value one = builder.create<arith::ConstantIndexOp>(loc, 1);
+  ReifiedRankedShapedTypeDims resultShape;
+  (void)reifyResultShapes(builder, resultShape);
+  for (auto dim : llvm::seq<int64_t>(0, outputRank)) {
+    loopBounds[dim].offset = zero;
+    loopBounds[dim].stride = one;
+    loopBounds[dim].size = resultShape[0][dim];
+  }
+  return loopBounds;
+}
+
+// Return the `interchangeVector` based on `dims_pos`.
+SmallVector<int64_t> computeInterchangeFromDimPos(ArrayRef<int64_t> dimsPos,
+                                                  int64_t inputRank) {
+  SmallVector<int64_t> interchangeVector;
+  interchangeVector.reserve(dimsPos.size());
+  // First map dims and their position. For example, dims_pos = [2, 0] will map
+  // to:
+  // [
+  //  [ key: 2, value: 0]
+  //  [ key: 0, value: 1]
+  // ]
+  // where key is the idx in dims_pos while value its position in dims_pos.
+  DenseMap<int64_t, int64_t> dimsAndPosMapping;
+  for (int64_t dimsIdx = 0, end = dimsPos.size(); dimsIdx < end; dimsIdx++)
+    dimsAndPosMapping[dimsPos[dimsIdx]] = dimsIdx;
+
+  // Scan the position in order and insert the value in the map
+  // to compute the interchange vector.
+  for (int64_t dimsIdx = 0; dimsIdx < inputRank; dimsIdx++) {
+    if (dimsAndPosMapping.count(dimsIdx))
+      interchangeVector.push_back(dimsAndPosMapping[dimsIdx]);
+  }
+  return interchangeVector;
+}
+
+// Implements `getIterationDomain` from the tiling interface.
+LogicalResult PackOp::generateScalarImplementation(OpBuilder &builder,
```

Comment:
I think if `getLoopIteratorType` returning just as many parallel loops as the input tensor type will make this change to have the `scf.for` for the tiled dimensions of the output being generated explicitly. This doesnt have to be part of this PR. It can be done after the fact (and I am happy to pick it up as well).
This way of doing it will make it compatible with implementing a tiling algorithm for this operation as well.